src/liger_kernel/ops/rms_norm.py

```diff
@@ -103,35 +126,78 @@ def _rms_norm_backward(
     dY_row = tl.load(dY_ptr + col_offsets, mask=mask, other=0)
     X_row = tl.load(X_ptr + col_offsets, mask=mask, other=0)
     W_row = tl.load(W_ptr + col_offsets, mask=mask, other=0)
+    original_x_dtype = X_row.dtype
 
     # Get cached rms
     inv_rms_row = tl.load(r_ptr)
 
     W_row = W_row + offset
-    dX_row = (inv_rms_row) * (
-        dY_row * W_row
-        - (1 / n_cols)
-        * inv_rms_row
-        * inv_rms_row
-        * tl.sum(dY_row * W_row * X_row, axis=0)
-        * X_row
-    )
-    tl.store(dY_ptr + col_offsets, dX_row, mask=mask)
+
+    # Different bacward graphs for different casting modes
+    if casting_mode == CASTING_MODE_LLAMA:
+        X_row = X_row.to(tl.float32)
+        m = (dY_row * W_row).to(tl.float32)
+        dX_row = inv_rms_row * m
+
+        dX_row += (inv_rms_row) * (
+            -(1 / n_cols)
+            * inv_rms_row
+            * inv_rms_row
+            * tl.sum(m * X_row, axis=0)
+            * X_row
+        )
+
+    if casting_mode == CASTING_MODE_GEMMA:
```

Comment:
looks like this only works with triton 3.0.0. If I reinstalled 2.3.0 triton can't compile  🤔 
```
E           triton.compiler.errors.CompilationError: at 38:23:    mask = col_offsets < n_cols
E           
E               Y_ptr += row_idx * Y_row_stride
E               X_ptr += row_idx * X_row_stride
E               r_ptr += row_idx * r_row_stride
E           
E               X_row = tl.load(X_ptr + col_offsets, mask=mask, other=0)
E               X_row_dtype = X_row.dtype
E               W_row = tl.load(W_ptr + col_offsets, mask=mask, other=0)
E           
E               # On Llama, only inv_rms is computed on fp32
E               if casting_mode == CASTING_MODE_LLAMA:
E                                  ^
E           AttributeError("'int' object has no attribute 'value'")

../.local/lib/python3.10/site-packages/triton/compiler/code_generator.py:1231: CompilationError
```

Reply:
@davidgonmar I made an attempt to fix this directly committing to your branch. The reason is when using triton < 3.0, the compiler is not smart enough to automatically treat global int var outside of jit function as a constexpr so `CASTING_MODE_XXX` remain int here. I explicitly convert the variables to `tl.constexpr` but making sure we're just passing the int value of it into the `jit`-ed function (otherwise the function signature will contain `tl.constexpr` arguments making it un-cachable)

Reply:
> @davidgonmar I made an attempt to fix this directly committing to your branch. The cause is when using triton < 3.0, the compiler is not smart enough to automatically treat global int var outside of jit function as a constexpr so `CASTING_MODE_XXX` remain int here. I explicitly convert the variables to `tl.constexpr` but making sure we're just passing the int value of it into the `jit`-ed function (otherwise the function signature will contain `tl.constexpr` arguments making it un-cachable) @lancerts @ByronHsu jfyi

thanks! sorry, did not know this did not work with older triton